File: simulator/src/piwarssim/engine/message/ClientScreenshotMessage.py
from piwarssim.engine.message.Message import Message


class ClientScreenshotMessage(Message):
    def __init__(self, factory, message_type):
        super(ClientScreenshotMessage, self).__init__(factory, message_type)
        self._frame_no = 0
        self._packet_no = 0
        self._total_packets = 0
        self._buffer = None

    def get_frame_no(self):
        return self._frame_no

    def get_packet_no(self):
        return self._packet_no

    def get_total_packets(self):
        return self._total_packets

    def get_buffer(self):
        return self._buffer

    def size(self):
        return super(ClientScreenshotMessage, self).size() + 4

    def deserialize_impl(self, deserializer):
        # super(FrameMessage, self).deserialize_impl(deserializer)
        self._frame_no = deserializer.deserialize_int()
        self._packet_no = deserializer.deserialize_unsigned_short()
        self._total_packets = deserializer.deserialize_unsigned_short()
        self._buffer = deserializer.deserialize_byte_array_raw()

    def serialize_impl(self, serializer):
        # super(FrameMessage, self).serialize_impl(serializer)
        pass # Not implemented

File: server/gcc-common/src/main/java/org/ah/gcc/virtualrover/message/ClientScreenshotMessage.java
package org.ah.gcc.virtualrover.message;

import org.ah.themvsus.engine.common.message.Message;
import org.ah.themvsus.engine.common.message.MessageCode;
import org.ah.themvsus.engine.common.message.MessageFactory;
import org.ah.themvsus.engine.common.transfer.Serializer;

import static org.ah.gcc.virtualrover.message.GCCMessageCode.ClientScreenshot;

public class ClientScreenshotMessage extends Message {

    private static final byte[] EMPTY_BUFFER = new byte[0];
    private int frameNo;
    private int packetNo;
    private int totalPackets;
    private byte[] buffer;
    private int off;
    private int len = 0;

    public ClientScreenshotMessage(MessageFactory messageFactory) {
        super(messageFactory);
    }

    public void setBuffer(byte[] buffer, int off, int len) {
        this.buffer = buffer;
        this.off = off;
        this.len = len;
    }

    @Override
    public void free() {
        super.free();
        len = 0;
        off = 0;
        buffer = null;
    }

    @Override
    public MessageCode getType() { return ClientScreenshot; }

    public void setFrameNo(int frameNo) {
        this.frameNo = frameNo;
    }

    public int getFrameNo() {
        return frameNo;
    }

    public void setPacketNo(int packetNo) {
        this.packetNo = packetNo;
    }

    public int getPacketNo() {
        return packetNo;
    }

    public void setTotalPackets(int totalPackets) {
        this.totalPackets = totalPackets;
    }

    public int getTotalPackets() {
        return totalPackets;
    }

    public int noDataSize() {
        return super.size() + 4 + 2 + 2 + 2;
    }

    @Override
    public int size() {
        return noDataSize() + len;
    }

    @Override
    protected void deserializeImpl(Serializer deserializer) {
        super.deserializeImpl(deserializer);
        frameNo = deserializer.deserializeInt();
        packetNo = deserializer.deserializeUnsignedShort();
        totalPackets = deserializer.deserializeUnsignedShort();
        buffer = deserializer.deserializeByteArray();
        off = 0;
        len = buffer.length;
    }

    @Override
    protected void serializeImpl(Serializer serializer) {
        serializer.serializeInt(frameNo);

        if (buffer == null) {
            serializer.serializeUnsignedShort(0);
            serializer.serializeUnsignedShort(0);
            serializer.serializeByteArray(EMPTY_BUFFER, 0, 0);
        } else {
            serializer.serializeUnsignedShort(packetNo);
            serializer.serializeUnsignedShort(totalPackets);
            serializer.serializeByteArray(buffer, off, len);
        }
    }
}
